models/multiple_solution/swarm_based/FHO.py

```diff
@@ -0,0 +1,187 @@
+            if len(preys_left)==0:
+                break
+        if len(preys_left)>0:
+            territories[len(Fhawks)-1]=np.array(list(territories[len(Fhawks)-1])+list(preys_left))
+        return territories
+            
+                      
+    def minimize_FHO(self):
+        ## Fire hawk algorithm to minimize the cost function       
+        
+        d=self.d
+        Xcandidates=self.Xcandidates
+        N=self.N
+        max_generations = self.max_generations  
+        minD=self.minD
+        maxD=self.maxD
+        Costfunction = self.cost_function
+        
+        #Evaluate the cost function for all candidate vectors
+        Cost= np.array([Costfunction(Xcandidates[i]) for i in range(N)])
+        
+        #Randomly set a number of Hawks between 1 and 20% of N
+        num_Hawks = np.random.randint(1,int(N/5)+1) if 1<int(N/5)+1 else 1
+        
+        #Ordering candidates
+        Xcandidates = Xcandidates[np.argsort(Cost)]
+        Cost.sort()
+        SP=Xcandidates.mean(axis=0)
+         
+        
+        #Select fire hawks
+        Fhawks= Xcandidates[:num_Hawks]
+        
+        #Select the Preys dim(N-num_Hawks,d)
+        Preys = Xcandidates[num_Hawks:]
+        
+        #get territories
+        territories=self.territories(Fhawks,Preys)
+        
+        #update best
+        GB=Cost[0]
+        Best_Hawk=Xcandidates[0]
+
+        #Counter
+        FEs=N
+
+        ## Main Loop
+        while FEs < max_generations:
+            PopTot=[]
+            Cost=[]
+            #Movement of Fire Hawk for all territories
+            for i in territories:
+                PR=territories[i].copy()
+                FHl=Fhawks[i].copy()
+                SPl=PR.mean(axis=0) if len(territories[i]) > 0 else np.zeros(FHl.shape)
+                a,b=np.random.uniform(0,1,size=2)
+                FHnear  =Fhawks[np.random.randint(num_Hawks)]                 
+                FHl_new =FHl+(a*GB-b*FHnear)
+                FHl_new = np.maximum(FHl_new,minD)
+                FHl_new = np.minimum(FHl_new,maxD)
+                PopTot.append(list(FHl_new))
+                
+                #Movement of the preys following Fire Hawks movement
+                for q in range(len(PR)): 
+                    a,b=np.random.uniform(0,1,size=2)
+                    PRq_new1=PR[q].copy()+((a*FHl-b*SPl))
+                    PRq_new1= np.maximum(PRq_new1,minD)
+                    PRq_new1 = np.minimum(PRq_new1,maxD)
+                    PopTot.append(list(PRq_new1))
+                    
+                    #Movement of the preys outside of territory
+                    a,b      =np.random.uniform(0,1,size=2)
+                    FHAlter  =Fhawks[np.random.randint(num_Hawks)] 
+                    PRq_new2 =PR[q].copy()+((a*FHAlter-b*SP));
+                    PRq_new2 = np.maximum(PRq_new2,minD)
+                    #The following line for PRq_new2 differs from original algorithm in matlab code (max instead of min):
```

Comment:
Just implement the equations as in the original paper. Many people they are faking the results of paper in the Matlab code. If you change the equations by your opinion, it is not their algorithm anymore. Rename this class as OriginalFHO and implement it as exactly same as in the original paper. You can create another class in the file name DevFHO (DevelopedFHO), based on your opinion. Compare two versions and see the results.

Reply:
Following your comment, two classes have been created in the FHO.py file :
-> Developed_FHO (including our modification)
-> Original_FHO (which sticks to the matlab code of the author).

Please note that the modified line is not explicitly covered in the article and may be a typo from the authors....

In all our tests (see notebook "FHO_VF.ipynb"), our algorithm converged to one of the global solutions... While the 'original' does not converge in 25-D for the Sphere function...